exceptions/designators_cache.go

```diff
@@ -0,0 +1,103 @@
+package exceptions
+
+import (
+	"hash/maphash"
+	"sync"
+
+	"github.com/armosec/armoapi-go/armotypes"
+)
+
+var (
+	globalDesignatorCache *designatorCache
+	setGlobalCacheOnce    sync.Once
+)
+
+type (
+	// designable knows how to identify the attributes from a Designator.
+	//
+	// TODO(fred): perf - this interface circumvents the unexported nature
+	// of armotypes.attributesDesignator.
+	// We could avoid this extra level of indirection by exporting that type.
+	designable interface {
+		GetCluster() string
+		GetNamespace() string
+		GetKind() string
+		GetName() string
+		GetPath() string
+		GetLabels() map[string]string
+	}
+
+	// designatorCache knows how to cache designators.
+	designatorCache struct {
+		mx       sync.RWMutex
+		innerMap map[portalDesignatorKey]designable
+		seed     maphash.Seed
+	}
+
+	portalDesignatorKey struct {
+		DesignatorType armotypes.DesignatorType
+		WLID           string
+		WildWLID       string
+		SID            string
+		AttributesHash uint64 // summarizes the map[string]string with a single hash value
+	}
+)
+
+// newDesignatorCache builds a chache for designators.
+//
+// This builds a single global instance only on the first call.
+func newDesignatorCache() *designatorCache {
+	setGlobalCacheOnce.Do(func() {
+		globalDesignatorCache = &designatorCache{
+			innerMap: make(map[portalDesignatorKey]designable, 1000),
+			seed:     maphash.MakeSeed(),
+		}
+	})
+
+	return globalDesignatorCache
+}
+
+func (c *designatorCache) Get(designator *armotypes.PortalDesignator) (designable, bool) {
+	c.mx.RLock()
+	defer c.mx.RUnlock()
+
+	val, ok := c.innerMap[c.toDesignatorKey(designator)]
+
+	return val, ok
+}
+
+func (c *designatorCache) Set(designator *armotypes.PortalDesignator, value designable) {
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	c.innerMap[c.toDesignatorKey(designator)] = value
+}
+
+func (c *designatorCache) toDesignatorKey(designator *armotypes.PortalDesignator) portalDesignatorKey {
+	return portalDesignatorKey{
+		DesignatorType: designator.DesignatorType,
+		WLID:           designator.WLID,
+		WildWLID:       designator.WildWLID,
+		SID:            designator.SID,
+		AttributesHash: c.hashMap(designator.Attributes), // a unique hash on attributes in order to make the key indexable
+	}
+}
+
+func (c *designatorCache) hashMap(input map[string]string) uint64 {
+	var (
+		sum uint64
+		h   maphash.Hash
+	)
+
+	for k, v := range input {
+		// NOTE: the hash is case sensitive. This shouldn't have any significant impact on performances
+		h.WriteString(k)
+		h.WriteByte(0)
+		h.WriteString(v)
+
+		// final hash has to be insensitive to the ordering of the map
+		sum += h.Sum64()
```

Comment:
did you consider using https://pkg.go.dev/hash/maphash#String here? see this discussion: https://github.com/golang/go/issues/42710

Reply:
no I missed that one... I started with fnv, but couldn't use that one without extra allocs, hence the hash/maphash...
Good hint.

Overall, this hashing is quite of a hassle, and we would just be better off if the WLID/WildWLID ids where somewhat guaranteed. At this moment, this is an information I don't have.

Reply:
@matthyx I've been considering your proposal, benchmarked it, then eventually reverted back to my initial choice of FNV, which is about 30% faster.

The current state of the PR is very draftish, as I've left tracks of the maphash implementation (with your piece of advice included), benchmark and extra profiling tests. I've also added a few tests there, since my initial version did mess up a bit.

Reply:
Interesting... so it's always better to measure than guess.
So I need to wait before having a look?

Reply:
@matthyx you may have a look now. The draftish status is intentional as there are still are lot of aspects yet to be decided.

To wit, the proposal in #75.
I wanted to be transparent about the core stuff being developed.
When we reach an agreement on the final shape of the exceptions package, I can move forward with more testing and the transitory/experimental code slashed.

Reply:
awesome, I'll have a look during lunch break

Reply:
@fredbi the proposal in #75 leads to catastrophic heap allocations, it's like 2x slower than on master
I need to check your assumption regarding the tuple so see if we can skip the hash calculation

Reply:
@matthyx ok I found the problem, my attempt to revert from global cache to privately-held cache overlooked the fact that we are creating this object in a loop. Fix upcoming soon

Reply:
@matthyx I've fixed this. My experimental branch for kubescape is now updated with the fix: https://github.com/kubescape/kubescape/pull/1054

Now the caches are being reused when iterating over results.

Reply:
indeed... so now I can review #75?

Reply:
yes. Please focus on the changes in the interface of the package. Once we agree on that part (this is a breaking change of API), I'll make a final round to trim away the profiling/experimental code.